python-frontend/src/main/java/org/sonar/python/types/InferredTypes.java

```diff
@@ -364,14 +365,26 @@ static boolean areSymbolsCompatible(Symbol actual, Symbol expected) {
     boolean isDuckTypeCompatible = !"NoneType".equals(otherFullyQualifiedName) &&
       expectedTypeClass.declaredMembers().stream().allMatch(m -> actualTypeClass.resolveMember(m.name()).isPresent());
     boolean canBeOrExtend = otherFullyQualifiedName == null || actualTypeClass.canBeOrExtend(otherFullyQualifiedName);
-    return areHardcodedCompatible || isDuckTypeCompatible || canBeOrExtend;
+    boolean areHardcodedIncompatible = areHardCodedIncompatible(actualTypeClass, expectedTypeClass);
+    return (areHardcodedCompatible || isDuckTypeCompatible || canBeOrExtend) && !areHardcodedIncompatible;
   }
 
   private static boolean areHardcodedCompatible(ClassSymbol actual, ClassSymbol expected) {
     Set<String> compatibleTypes = HARDCODED_COMPATIBLE_TYPES.getOrDefault(actual.fullyQualifiedName(), Collections.emptySet());
     return compatibleTypes.stream().anyMatch(expected::canBeOrExtend);
   }
 
+  private static boolean areHardCodedIncompatible(ClassSymbol actual, ClassSymbol expected) {
```

Comment:
what if we'll need to support other sets of hardcoded incompatible types?
I think a solution like below could cover it and be a bit cleaner
```java
private static final Set<Set<String>> INCOMPATIBLE_TYPES = Set.of(Set.of("tuple", "list"));

  private static boolean areHardCodedIncompatible(ClassSymbol actual, ClassSymbol expected) {
    return INCOMPATIBLE_TYPES.contains(
      Stream.of(actual, expected)
        .map(ClassSymbol::fullyQualifiedName)
        .filter(Objects::nonNull)
        .collect(Collectors.toSet())
    );
  }
```

Reply:
I honestly hope we won't have to use this trick for other incompatible types, but I agree this approach looks better, thanks!